atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java

```diff
@@ -0,0 +1,168 @@
+ * InputStream, write the uncompressed data through the OutputStream, then
+ * serve the read from the resulting compressed buffer.
+ */
+public final class LZ4CompressingInputStream extends BufferedDelegateInputStream {
+
+    private static final LZ4Compressor COMPRESSOR = LZ4Factory.fastestInstance().fastCompressor();
+    private static final int DEFAULT_SEED = 0x9747b28c;
+    private static final int DEFAULT_BLOCK_SIZE = 1 << 16; // 64 KB
+
+    private final LZ4BlockOutputStream compressingStream;
+    private final int blockSize;
+
+    // Flag to prevent writes to the internal buffer while the
+    // compressing stream is reading from it.
+    private boolean readInProgress;
+    // Flag to indicate whether this stream has been exhausted.
+    private boolean finished;
+
+    public LZ4CompressingInputStream(InputStream delegate) throws IOException {
+        this(delegate, DEFAULT_BLOCK_SIZE);
+    }
+
+    public LZ4CompressingInputStream(InputStream delegate, int blockSize) throws IOException {
+        super(delegate, Math.max(blockSize, COMPRESSOR.maxCompressedLength(blockSize)));
+        this.blockSize = blockSize;
+        Checksum checksum = XXHashFactory.fastestInstance().newStreamingHash32(DEFAULT_SEED).asChecksum();
+        OutputStream delegateOutputStream = new InternalByteArrayOutputStream();
+        this.compressingStream = new LZ4BlockOutputStream(delegateOutputStream, blockSize, COMPRESSOR, checksum, true);
+        this.readInProgress = false;
+        this.finished = false;
+
+        // Flush the LZ4 header so that the compression buffer is empty
+        delegateOutputStream.flush();
+    }
+
+    @Override
+    protected boolean refill() throws IOException {
+        if (finished) {
+            return false;
+        }
+        position = 0;
+        bufferSize = 0;
+
+        // Read a block of data from the delegate input stream.
+        int bytesRead = ByteStreams.read(delegate, buffer, BUFFER_START, blockSize);
+        if (bytesRead == 0) {
+            // The delegate input stream has been exhausted, so we notify
+            // the compressing stream that there are no remaining bytes.
+            // Lastly, we flush the LZ4 footer into the internal buffer.
+            compressingStream.finish();
+            compressingStream.flush();
+            finished = true;
+            return true;
+        }
+
+        // Write the bytes to the compressing stream and flush it.
+        // The compressing stream flushes bytes when it contains more
+        // than blockSize bytes, so we manually flush after writing up to
+        // blockSize bytes to flush the compressed bytes to the internal
+        // buffer. The readInProgress flag protects against simultaneously
+        // reading and writing to the same buffer. The write call will read
+        // from the buffer but should not write to it - only the flush call
+        // should.
+        readInProgress = true;
```

Comment:
might be cleaner to use a separate buffers for reading and writing, so we don't have to worry about this